Add end-to-end test for pj-rehearse

petr-muller opened the pull request:
Craft an `openshift/release`-like repo, as if prepared by cloneref.
Prepare a `JOB_SPEC` to instruct pj-rehearse to work over that repo (in
dry mode), and roughly validate its output (specific content of rehearsal
jobs is validated on unit test level).

In a separate commit, I have also adjusted integration test directories,
naming and `Makefile` targets to the fact that we now have multiple things
to test.

I have reused @droslean 's test data from https://github.com/openshift/ci-operator-prowgen/pull/67

/cc @stevekuznetsov @droslean @bbguimaraes

petr-muller committed:
Add end-to-end test for pj-rehearse

Craft an `openshift/release`-like repo, as if prepared by cloneref.
Prepare a `JOB_SPEC` to instruct pj-rehearse to work over that repo (in
dry mode), and roughly validate its output.

petr-muller committed:
Make integration tests structure & naming consistent

petr-muller commented:
grr stupid old git

stevekuznetsov commented on test/pj-rehearse-integration/run.sh:
```diff
@@ -0,0 +1,54 @@
+echo "[INFO] Running pj-rehearse in dry-mode..."
+submitted_yamls="${workdir}/rehearsals.yaml"
+error_log="${workdir}/errors.log"
+export JOB_SPEC='{"type":"presubmit","job":"pull-ci-openshift-release-master-rehearse","buildid":"0","prowjobid":"uuid","refs":{"org":"openshift","repo":"release","base_ref":"master","base_sha":"'${base_sha}'","pulls":[{"number":1234,"author":"petr-muller","sha":"'${candidate_sha}'"}]}}'
```
Oh boy

petr-muller replied:
Maybe I'll extract it to a function so it does not clutter the flow that much.

stevekuznetsov commented on test/pj-rehearse-integration/run.sh:
```diff
@@ -0,0 +1,54 @@
+
+echo "[INFO] Validating created rehearsals"
+rehearsed_jobs="${workdir}/rehearsed_jobs"
+grep "^  job: rehearse-1234-pull-ci" "${submitted_yamls}" | cut -c '8-' | sort  > "${rehearsed_jobs}"
```
While this is like 70% of what we want, could we invent a different type of dry-run mode or something that will either POST to a fake server we have or just spit out the YAML it would have POSTed so we can validate like 100% of the output here?

petr-muller replied:
I considered it but decided not to because the lower-level details are tested in lower level tests. But I'm willing to negotiate :)

Dry-run is certainly worth fixing so it outputs a proper YAML (list of jobs instead of concatenated individual job YAMLs) so it can be parsed back - that's quite easy. We could then write a Python helper or something to validate it against some expected output, but we would need to deal with the dynamic fields so it's not ultra trivial, and I'm not sure it's worth it because we test the job->rehearsal translation on unit level.

stevekuznetsov replied:
Unit tests can't replace e2e! I think we should try -- dynamic fields can be elided with `diff --ignore-matching-lines`, we manage that with k8s resources often. Plus, if we're not actually talking to a server we won't be getting fields like `creationTimestamp` so we should have fewer dynamic fields.

petr-muller replied:
Not replace but complement, which I think is the case here. But OK, I'll give it a shot.

petr-muller committed:
pj-rehearse: change how dry-mode outputs YAML

Previously, the dry client was responsible for printing out YAML
marshals of the jobs that would be submitted. This resulted in
concatenated YAML objects printed to output. To provide parseable
output, it needs to be printed as a YAML list instead of the
concatenation.

petr-muller commented:
@stevekuznetsov PTAL now. I have iterated a bit on the shell script so it is slightly more legible. It now does a full diff over the output YAML, excluding some lines (but hell, that basic regex syntax...). Obtaining deterministic output also needed a change in how the YAML is actually produced by `pj-rehearse`.

stevekuznetsov commented on pkg/rehearse/jobs.go:
```diff
@@ -301,9 +308,11 @@ func (e *Executor) waitForJobs(jobs sets.String, selector string) (bool, error)
+func (e *Executor) submitRehearsals() (sets.String, []*pjapi.ProwJob, error) {
 	var errors []error
-	pjs := make(sets.String, len(e.rehearsals))
+	pj_names := make(sets.String, len(e.rehearsals))
```
Prefer `pjNames` over `pj_names`

stevekuznetsov replied:
I know you didn't change this, but `len(e.rehearsals)` is the worst case scenario for the size of this set, right? Why do we need to do something other than use the `sets.NewStrings()` constructor?

bbguimaraes replied:
It is the "worst case scenario" in the sense that it will not be entirely filled if errors occur when creating the `PJ`s, which is an exceptional case.

In all other cases the set will be exactly this size, hence the pre-allocation.

petr-muller replied:
Addressed.

stevekuznetsov commented on pkg/rehearse/jobs.go:
```diff
@@ -230,19 +225,31 @@ func NewExecutor(toBeRehearsed map[string][]prowconfig.Presubmit, prNumber int,
 func (e *Executor) ExecuteJobs() (bool, error) {
 	submitSuccess := true
-	pjs, err := e.submitRehearsals()
+	names, pjs, err := e.submitRehearsals()
```
If you change to just return the prowjob structs, can you do the reduction to a set of strings here rather than in the `submitRehearsals()`?

petr-muller replied:
Addressed.

stevekuznetsov commented on test/pj-rehearse-integration/run.sh:
```diff
@@ -0,0 +1,64 @@
+fi
+
+echo "[INFO] Validating created rehearsals"
+if ! compare_to_expected "$REHEARSED_JOBS" > "${WORKDIR}/diff"; then
```
FYI -- not an issue right now but the `if ! function_call` structure will remove your ability to influence control flow through the script via `set -o errexit` in the function call. We're just calling `diff` once so it's not an issue right now

stevekuznetsov replied:
You can also capture the output without an intermediate file if you want:

```sh
if ! output="$( compare_to_expected "$REHEARSED_JOBS" )"; then
```

stevekuznetsov replied:
nit: `"${REHEARSED_JOBS}"`

petr-muller replied:
Addressed. I wanted to hide the diff params in the function. I really hope the script will not really grow beyond this.

stevekuznetsov commented on test/pj-rehearse-integration/expected_jobs.yaml:
```diff
@@ -0,0 +1,398 @@
+- apiVersion: prow.k8s.io/v1
```
How come `pj-rehearse` doesn't have the normal user-facing output?

bbguimaraes replied:
I was also wondering that. We could add a flag similar to `oc`'s `-o yaml` — which I imagine could be useful in other contexts ­— to avoid sacrificing the valuable default output, akin to what we did for `--debug-log`.

petr-muller replied:
It only has logging ATM... Maybe we could fix it but right now clear stdout is quite handy in dry-mode :)

stevekuznetsov commented on test/pj-rehearse-integration/run.sh:
```diff
@@ -0,0 +1,64 @@
+echo "[INFO] Validating created rehearsals"
+if ! compare_to_expected "$REHEARSED_JOBS" > "${WORKDIR}/diff"; then
+  echo "[ERROR] pj-rehearse output differs from expected:"
+  cat "${WORKDIR}/diff"
```
nit: prepending these with `[ERROR]` will get them highlighted in gubernator but if no other scripts in the repo are doing that then forget it

petr-muller replied:
Addressed.

bbguimaraes commented on pkg/rehearse/jobs.go:
```diff
@@ -44,12 +45,6 @@ func NewProwJobClient(clusterConfig *rest.Config, namespace string, dry bool) (p
-			}
-			fmt.Printf("%s\n", jobAsYAML)
 			return false, nil, nil
 		})
```
The reactor doesn't serve any purpose anymore, it can be removed.

petr-muller replied:
Addressed. Also gets rid of two imports which is nice.

bbguimaraes commented on test/pj-rehearse-integration/run.sh:
```diff
@@ -0,0 +1,64 @@
+set -o nounset
+set -o pipefail
+
+readonly WORKDIR="$( mktemp -d )"
```
_it looks like @stevekuznetsov forgot the -Wpedantic flag_

`readonly` will mask the result of `mktemp` in case of failure.

petr-muller replied:
Addressed. Apparently `readonly` works slightly differently than I expected. I actually rarely used it before. That's what you get for trying :/

bbguimaraes commented on test/pj-rehearse-integration/run.sh:
```diff
@@ -0,0 +1,64 @@
+  mkdir "${FAKE_OPENSHIFT_RELEASE}"
+  pushd "${FAKE_OPENSHIFT_RELEASE}" >/dev/null
+  git init --quiet
+  cp -R "${master_data}"/* .
```
I suspect `git clone`ing the original repository would be faster.

petr-muller replied:
I don't think we need to care about speed here, it's just a few files.

bbguimaraes commented on test/pj-rehearse-integration/run.sh:
```diff
@@ -0,0 +1,64 @@
+  cp -R "${master_data}"/* .
+  git add ci-operator cluster
+  git commit -m "Master version of openshift/release" --quiet
+  local -r base_sha="$(git rev-parse HEAD)"
```
Same as `readonly` above.

petr-muller replied:
Addressed.

petr-muller commented:
/hold

Addressed the feedback in a separate commit for clarity. I'll squash before merge.

stevekuznetsov commented:
/lgtm
Squash away

petr-muller commented:
/hold cancel

I have no idea why the `lgtm` is gone, I just squashed last two commits :(

droslean commented:
/lgtm
